core/service-test-runner/infer-pull-request.js

```diff
@@ -64,13 +64,13 @@ function _inferPullRequestFromCircleEnv(env) {
  * When called inside a CI build, infer the details
  * of a pull request from the environment variables.
  *
- * @param {object} [env=process.env]
- * @return {object} pr
- * @return {string} pr.baseUrl (returned for travis CI only)
- * @return {string} pr.owner
- * @return {string} pr.repo
- * @return {string} pr.pullRequest PR/issue number
- * @return {string} pr.slug owner/repo/#pullRequest
+ * @param {object} [env=process.env] Environment variables
+ * @returns {object} pr
+ * @returns {string} pr.baseUrl (returned for travis CI only)
+ * @returns {string} pr.owner
+ * @returns {string} pr.repo
+ * @returns {string} pr.pullRequest PR/issue number
+ * @returns {string} pr.slug owner/repo/#pullRequest
```

Comment:
I'm not really sure what to do about these. Its useful to be able to describe the structure of a soft object. This is quite human-readable and some JSDoc --> HTML/markdown/whatevz tools do accept this multiple return syntax and render it validly. For example, doxdox is perfectly happy to render this: https://doxdox.org/badges/shields#core-service-test-runner-infer-pull-request.js-inferpullrequest

That said:

* doxdox does have some other issues - I've just picked it here because it can render docs straight from github which is convenient
* some tools do only render the first/last return

..but doing this throws `Found more than one @returns declaration` in the linter. The other options are:

* `@returns {object} pr`, which is less helpful
* Convert it to a `@typedef` which does render in some tools, but also `@typedef` is not supported by some others and defining the types outside of the function definition is less human-redable/editor hinting-friendly.

There doesn't seem to be a great solution to this. I guess it kind of depends on which JSDoc --> HTML/markdown/whatevz tool we opt for (spoiler: every single one I've tried so far doesn't do something important we want).

The other thing that is annoying is that multiple returns isn't a single check. You have to disable both `require-returns` and `require-returns-check` to not throw it.

Reply:
A lint ignore is another option.

I can't say this particular helper function feels all that important, in the scheme of things. Are there many others of these? If it's just this function, I'm not sure it matters all that much which solution you pick.

Reply:
I picked this one to leave a comment on, but we do have several occurrences of this.
Its something we're going to run into as we document more stuff though. In general,

```js
return {
  foo: 73,
  bar: 'hello',
  baz: false,
}
```

is a very common thing to do in javascript and in any sort of meaningful docs, we're going to want to describe what `foo`, `bar` and `baz` are. As I say, I think it partly depends on what tooling we end up using so maybe that's the next sensible thing to look at before writing more docs.

Reply:
One final way to think about this may be to say that if we've got complex objects we're going to pass around that need describing, maybe we should formalise them as classes. I tend to be of the opinion that the purpose of documentation is to describe what we've done rather than to decide how we do it. That said, maybe there are some situations where the act of describing the code is also helping to see how we should improve it.

For the moment, I've just disabled the lint rules to get rid of the remaining warnings so we can merge this, but I'll continue to think about this.

Reply:
Another approach I've seen folks used in the past is defining a doc typedef for the return type:
```javascript
/**
 * @typedef {object} ThatReturnedThing
 * @property {number} foo - Lorem ipsum.
 * @property {string} bar - Lorem ipsum.
 * @property {boolean} baz - Lorem ipsum.
 */
```

and then using it on the function doc
```js
 * ...
 * @returns {ThatReturnedThing} 
 */
```

Reply:
Yep. I did mention `@typedef`, but support for it seems to be non-universal - hence next thing to do is think about tooling..
Out of interest, do you have any suggestions for JSDoc renderers worth having a look at?

Reply:
> Yep. I did mention @typedef

Sorry 🤦‍♂ And that's why it is important to read thoroughly!

> Out of interest, do you have any suggestions for JSDoc renderers worth having a look at?

Not really. I kinda like the output of Toast UI template (https://www.npmjs.com/package/tui-jsdoc-template) that I've seen used in a few places, like Yeoman https://yeoman.github.io/generator/, but I don't have much experience nor strong opinions with any of them.

Reply:
Have you looked at https://pypi.org/project/sphinx-js/ by chance? I'm not the hugest fan of Sphinx. I find it a pain to set up and I don't really know RsT. But I do use it; and once it's set up it works.

Reply:
No worries + cheers for the suggestions, both.

I think having tried out a few different tools, JSDoc3 (which could be used with the toast theme if we want) + typedefs is the right way to go, but I think JSDoc3 also requires some additional annotations for modules and stuff. I reckon we might as well aim to get it right in this PR because the tooling we adopt slightly affects how we want to do the annotations.

Leave this PR open. I'll expand the scope slightly to also adopting a renderer as well as some lint rules. I'll do some more work on it and push some more commits.

Reply:
Seems like we ought to deploy the developer docs separately from the shields.io app so they are always in sync with master. Ideally we'd get review apps for PRs as well. I've been playing with Netlify which is extraordinarily easy to set up and might be a nice option; would also keep it out of the standard Heroku build for self-hosting users.

Not sure if you want to pull in the hosting for this PR too. I'm happy to help with that in this branch if you'd like.

Reply:
Yes - agree that we should get docs building + deploying sooner rather than later as its definitely going to be useful to:

a) get them somewhere contributors can see/use them
b) check how they render in staging when reviewing PRs that contribute docs

Thanks for the offer of help. I haven't got as far as thinking about the best way to host it, so happy to defer to your ideas on this. Normally I'd bung the docs on the gh-pages branch, but we're already using that on this repo.

I think it would probably be better to do that in a separate PR though as this one is already touching a lot of files. 2 smaller diffs will be easier to review.